server/catgenome/src/main/java/com/epam/catgenome/util/feature/reader/TabixReader.java

```diff
@@ -0,0 +1,682 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import com.epam.catgenome.util.Utils;
+import htsjdk.samtools.seekablestream.ISeekableStreamFactory;
+import htsjdk.samtools.seekablestream.SeekableStream;
+import htsjdk.samtools.seekablestream.SeekableStreamFactory;
+import htsjdk.samtools.util.BlockCompressedInputStream;
+import htsjdk.tribble.Feature;
+import htsjdk.tribble.FeatureCodec;
+import htsjdk.tribble.FeatureCodecHeader;
+import htsjdk.tribble.TribbleException;
+import htsjdk.tribble.util.ParsingUtils;
+import htsjdk.tribble.util.TabixUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * @author Heng Li <hengli@broadinstitute.org>
+ */
+public class TabixReader {
+    private EhCacheBasedIndexCache indexCache;
+    private String mFn;
+    private String mIdxFn;
+    private BlockCompressedInputStream mFp;
+
+    private int mPreset;
+    private int mSc;
+    private int mBc;
+    private int mEc;
+    private int mMeta;
+    //private int mSkip; (not used)
+    private String[] mSeq;
+
+    private Map<String, Integer> mChr2tid;
+
+    private static int maxBin = 37450;
+    //private static int TAD_MIN_CHUNK_GAP = 32768; (not used)
+    private static int tadLidxShift = 14;
+
+    protected static class TPair64 implements Comparable<TPair64> {
+        long u, v;
+
+        public TPair64(final long _u, final long _v) {
+            u = _u;
+            v = _v;
+        }
+
+        public TPair64(final TPair64 p) {
+            u = p.u;
+            v = p.v;
+        }
+
+        public int compareTo(final TPair64 p) {
+            return u == p.u ? 0 : ((u < p.u) ^ (u < 0) ^ (p.u < 0)) ? -1 : 1; // unsigned 64-bit comparison
+        }
+    }
+
+    protected static class TIndex {
+        HashMap<Integer, TPair64[]> b; // binning index
+        long[] l; // linear index
+    }
+
+    protected TIndex[] mIndex;
+
+    protected static class TIndexCache <T extends Feature, S> implements CacheIndex {
+        TIndex[] mIndex;
+        String[] mSeq;
+        Map<String, Integer> mChr2tid;
+        int mPreset;
+        int mSc;
+        int mBc;
+        int mEc;
+        int mMeta;
+        FeatureCodecHeader header;
+        FeatureCodec <T, S> codec;
+        SeekableStream fp;
```

Comment:
Not sure if it is okay to store a SeekableStream in the cache: it needs to be closed sometime. Is it really needed to store this in the cache?

Reply:
Caching of stream is deleted.